### DISCUSSION TITLE
Best practices for passing large dictionaries to client components?

### DISCUSSION THREAD

**Discussion by juliusn** [amannn/next-intl#656] [Q&A]
Because translations are encouraged to keep on the server and pass to client components as props, what is the recommended workflow for passing large dictionaries? 

messages/en.json:
```json
{
  "Register": {
    "title": "Register new account",
    "firstName": "First name",
    "lastName": "Last name",
    "email": "Email",
    "emailInvalid": "Email is invalid.",
    "password": "Password",
    "passwordInvalid": "Password should be at least 8 characters.",
    "passwordAgain": "Password again",
    "passwordsDontMatch": "Passwords don't match",
    "submit": "Submit",
    "alreadyHaveAccount": "Already have an account? <link>Log in here!</link>"
  }
}
```

app/[locale]/register/page.tsx
```tsx
import { useTranslations } from 'next-intl';
import { RegisterForm } from './RegisterForm';
import Link from 'next/link';

export default function RegisterPage() {
  const t = useTranslations('Register');

  return (
    <>
      <h2>{t('title')}</h2>
      <RegisterForm
        firstName={t('firstName')}
        lastName={t('lastName')}
        email={t('email')}
        emailInvalid={t('emailInvalid')}
        password={t('password')}
        passwordInvalid={t('passwordInvalid')}
        passwordAgain={t('passwordAgain')}
        passwordsDontMatch={t('passwordsDontMatch')}
        submit={t('submit')}
      />
      <p>
        {t.rich('alreadyHaveAccount', {
          link: (text) => <Link href="/login">{text}</Link>,
        })}
      </p>
    </>
  );
}
```

app/[locale]/register/RegisterForm.tsx
```tsx
'use client';

type RegisterFormProps = {
  firstName: string;
  lastName: string;
  email: string;
  emailInvalid: string;
  password: string;
  passwordInvalid: string;
  passwordAgain: string;
  passwordsDontMatch: string;
  submit: string;
};

export function RegisterForm(props: RegisterFormProps) {
  return (
    <form>
      <label htmlFor="firstName">{props.firstName}</label>
      <input id="firstName" type="text" name="firstName" />
      <label htmlFor="lastName">{props.lastName}</label>
      <input id="lastName" type="text" name="lastName" />
      <label htmlFor="email">{props.email}</label>
      <input id="email" type="text" name="email" />
      {/* and so on... */}
    </form>
  );
}
```

Here's a typical example where I'm passing down 9 messages, which is quite a moderate amount considering that a client component often needs to handle things like form validation, error handling, displaying notifications and stuff like that. Since there doesn't seem to be an API method for receiving a library path as a dictionary object, is the only solution just to pass them one by one like this?

**Comment by mamlzy** [comment]
@amannn can you give us the best practice to do this?, thanks

**Comment by amannn** [comment]
@juliusn @ImamAlfariziSyahputra I think it depends a bit on the use case. Passing a few labels to a component might be fine, but if you find that you're passing a large number of messages, I'd consider if the component tree can be adapted to use more granular client components.

Here's an attempt for the register form use case:

```tsx
// register/page.tsx

// A Client Component, so that it can use:
// - `useFormStatus` to disable the input field while submitting
// - `useFormState` to display error messages received from server
import FormField from './FormField';

// A Client Component, so that it can use `useFormStatus`
// to disable the submit button while submitting
import FormSubmitButton from './FormSubmitButton';

export default function RegisterPage() {
  const t = useTranslations('RegisterPage');

  function onRegister() {
    'use server';
    // ...
  }

  return (
    <form action={onRegister}>
      <FormField label={t('firstName')} name="firstName" />
      <FormField label={t('lastName')} name="lastName" />
      <FormField error={t('emailInvalid')} label={t('email')} name="email" />
      <FormField
        error={t('passwordInvalid')}
        label={t('password')}
        name="password"
      />
      <FormField
        error={t('passwordAgainInvalid')}
        label={t('passwordAgain')}
        name="passwordAgain"
      />
      <FormField
        error={t('passwordInvalid')}
        label={t('password')}
        name="password"
      />
      <FormSubmitButton label={t('submit')} />
    </form>
  );
}
```

There are use cases where handling i18n on the server side becomes too cumbersome though if you need a lot of interactivity. Using `NextIntlClientProvider` for these cases is definitely fine.

Does this help?

**Reply by amannn** [reply]
I've added this as an example to https://next-intl-docs.vercel.app/docs/environments/server-client-components#option-1-passing-translations-to-client-components ("Example: How can I implement a form?") and cleaned up the code a bit.

**Reply by juliusn** [reply]
In a situation where the elements, such as form fields, aren't coupled with the container component, moving them to child components sounds like a viable solution. However, often an input element may consume state variables and event listeners that are defined in the parent form, so they need to be passed down as well. In cases like this there might not be any improvement in readibility, which is my main concern. 

I started another thread about [benefits of server-side internationalization](https://github.com/amannn/next-intl/discussions/660). In a situation where I'm using the `NextIntlClientProvider` anyway, what exactly am I sacrificing when moving translations into client-side? The documentation doesn't cover this in detail.

**Reply by amannn** [reply]
> However, often an input element may consume state variables and event listeners that are defined in the parent form, so they need to be passed down as well.

Sure, for such cases you need to handle i18n in the client-side code—that's totally fine.

How much handling i18n on the client side affects you depends on your app and your requirements (e.g. bundle size, TBT, etc.). It can be really fast though, I have an app, pre-App Router, with a large number of components using i18n on the client side and the performance of the app is great. There are also further performance improvements planned for handling i18n on the client side, so definitely go for it if it works better for your use case!
